crates/rattler_installs_packages/src/wheel_builder.rs

```diff
@@ -0,0 +1,370 @@
+    /// A cache for virtualenvs that might be reused later in the process
+    venv_cache: BuildCache,
+
+    /// The package database to use
+    package_db: &'db PackageDb,
+
+    /// The env markers to use when resolving
+    env_markers: &'i MarkerEnvironment,
+
+    /// The configured wheel tags to use when resolving
+    wheel_tags: Option<&'i WheelTags>,
+
+    /// The resolve options. Note that we change the sdist resolution to normal if it's set to
+    /// only sdists, because otherwise we run into a chicken & egg problem where a sdist is required
+    /// to build a sdist. E.g. `hatchling` requires `hatchling` as build system.
+    resolve_options: ResolveOptions,
+}
+
+/// An error that can occur while building a wheel
+#[allow(missing_docs)]
+#[derive(thiserror::Error, Debug)]
+pub enum WheelBuildError {
+    #[error("Could not build wheel: {0}")]
+    Error(String),
+
+    #[error("Could not install artifact in virtual environment")]
+    UnpackError(#[from] UnpackError),
+
+    #[error("Could not build wheel: {0}")]
+    IoError(#[from] std::io::Error),
+
+    #[error("Could not run command {0} to build wheel: {1}")]
+    CouldNotRunCommand(String, std::io::Error),
+
+    #[error("Could not resolve environment for wheel building")]
+    CouldNotResolveEnvironment(Vec<Requirement>),
+
+    #[error("Error parsing JSON from extra_requirements.json: {0}")]
+    JSONError(#[from] serde_json::Error),
+
+    #[error("Could not parse generated wheel metadata: {0}")]
+    WheelCoreMetadataError(#[from] WheelCoreMetaDataError),
+
+    #[error("Could not get artifact")]
+    CouldNotGetArtifact,
+}
+
+impl<'db, 'i> WheelBuilder<'db, 'i> {
+    /// Create a new wheel builder
+    #[must_use]
+    pub fn new(
+        package_db: &'db PackageDb,
+        env_markers: &'i MarkerEnvironment,
+        wheel_tags: Option<&'i WheelTags>,
+        resolve_options: &'i ResolveOptions,
+    ) -> Self {
+        // We are running into a chicken & egg problem if we want to build wheels for packages that
+        // require their build system as sdist as well. For example, `hatchling` requires `hatchling` as
+        // build system. Hypothetically we'd have to look through all the hatchling sdists to find the one
+        // that doesn't depend on itself.
+        // Instead, we use wheels to build wheels.
+        let resolve_options = if resolve_options.sdist_resolution == SDistResolution::OnlySDists {
+            ResolveOptions {
+                sdist_resolution: SDistResolution::Normal,
+            }
+        } else {
+            resolve_options.clone()
+        };
+
+        Self {
+            venv_cache: RefCell::new(HashMap::new()),
+            package_db,
+            env_markers,
+            wheel_tags,
+            resolve_options,
+        }
+    }
+
+    /// Get a prepared virtualenv for building a wheel (or extracting metadata) from an `[SDist]`
+    /// This function also caches the virtualenvs, so that they can be reused later.
```

Comment:
If we cache virtualenvs, do we want to do that over runs? Or just within a single resolve?

Reply:
within a single resolve IMO, because you might do multiple steps (prepare metadata, and then you can reuse the env to build the wheel etc.)